File: fsharp/grade-school/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = 
    Map.empty

let add (student: string) (grade: int) (school: School): School = 
    match (Map.containsKey grade school) with
    | true -> 
        let students = List.sort (student :: Map.find grade school)
        school.Add (grade, students)
    | false -> 
        school.Add (grade, [student])

let roster (school: School): string list = 
    Map.fold 
        (fun acc key value -> (acc @ value))
        []
        school

let grade (number: int) (school: School): string list = 
    match (Map.containsKey number school) with
    | true  -> Map.find number school
    | false -> []

File: csharp/grade-school/GradeSchool.cs
using System.Collections.Generic;
using System.Linq;

public class GradeSchool
{
    private SortedDictionary<int, SortedSet<string>> Grades { get; set; } = new();
    
    public bool Add(string student, int grade)
    {
        IEnumerable<string> names = Roster();
        if (names.Contains(student)) return false;
        if (!Grades.ContainsKey(grade)) Grades[grade] = [];

        Grades[grade].Add(student);
        return true;
    }

    public IEnumerable<string> Roster() => Grades.Values.SelectMany(name => name);

    public IEnumerable<string> Grade(int grade) => 
        Grades.TryGetValue(grade, out var namesInGrade) ? namesInGrade : new List<string>();
}
